planner/core/expression_rewriter.go

```diff
@@ -1175,6 +1178,25 @@ func (er *expressionRewriter) betweenToExpression(v *ast.BetweenExpr) {
 // Otherwise it should return false to indicate (the caller) that original behavior needs to be performed.
 func (er *expressionRewriter) rewriteFuncCall(v *ast.FuncCallExpr) bool {
 	switch v.FnName.L {
+	// when column is not null, ifnull on such column is not necessary.
+	case ast.Ifnull:
+		if len(v.Args) != 2 {
+			er.err = expression.ErrIncorrectParameterCount.GenWithStackByArgs(v.FnName.O)
+			return true
+		}
+		stackLen := len(er.ctxStack)
+		arg1 := er.ctxStack[stackLen-2]
+		newArg1, isColumn := arg1.(*expression.Column)
+
+		// if expr1 is a column and column has not null flag, then we can eliminate ifnull on
+		// this column.
+		if isColumn && mysql.HasNotNullFlag(newArg1.RetType.Flag) && !er.isOuterJoin {
```

Comment:
@zhexuany it's not correct, and a little tricky. 

We should:
1. maintain the `NotNull` flag on each column of the output schema for each operator.
2. eliminate the `IfNull` function if the first input parameter has `NotNull` flag.

The `!er.isOuterJoin` should be removed. It makes no sense, and a little redundant. Since you have considered the `Join` operator, why not considering the `Aggregate` operator? The aggregate function output can not be NULL as well.

Reply:
I know it's hard to maintain the `NotNull` flag for each operator. We can maintain the correct `NotNull` flag only for the join operator now:
1. if it's a inner join, and the output inner column is not in the join equal condition, the `NotNull` flag should be removed. In order to solve the data race problem, the result type might also be copied.
2. if it's a outer join, all the output inner columns should not have the `NotNull` flag.

With the above rules, we can promise that the `NotNull` flag is correct. The remained task is to propagate the `NotNull` flags in other operators and functions.

Reply:
@zhexuany As I described in the above:
>  In order to solve the data race problem, the result type might also be copied.

Reply:
It is kinda ambiguous to me.  

> the result type might also be copied.

Could I interpret above to "change `RetType from reference to value`"?

Reply:
If you want to change the `flag` of the return type, it's better to make a copy first. Otherwise, the copy step can be saved.